### PR TITLE
apm-data: add `event.ingested` field to APM documents

### PR DESCRIPTION
**Pull request by inge4pres**
### APM Data plugin

Due to migration in 8.15 to the plugin, not using anymore the Fleet asset management in integrations, the `event.ingested` is not populated.

Add the field again in our generally shared mappings and ingest pipeline to restore functionalities used by other products on APM documents.

### Checks

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)? ✅ 
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md)? ✅ 
- If submitting code, is your pull request against main? Unless there is a good reason otherwise, we prefer pull requests against main and will backport as needed. ✅

### CODE REVIEW

File: x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml [hunk 1]
```diff
@@ -24,3 +24,13 @@ processors:
...
+- script:
+    description: Add time when event was ingested (and remove sub-seconds to improve storage efficiency)
+    tag: truncate-subseconds-event-ingested
+    ignore_failure: true
+    source: |-
+      if (ctx?.event == null) {
+        ctx.event = [:];
+      }
+
+      ctx.event.ingested = metadata().now.withNano(0).format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
```
Comment (inge4pres) [inline x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:36]: Is there a way to _not_ set the field if it's already present?

File: x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml [hunk 2]
```diff
@@ -24,3 +24,13 @@ processors:
     target_field: user_agent
     ignore_failure: true
     ignore_missing: true
+- script:
+    description: Add time when event was ingested (and remove sub-seconds to improve storage efficiency)
```
Comment (inge4pres) [inline x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:28]: I admit i copied this from Fleet.
Do we want to adopt the same second-granularity?

File: x-pack/plugin/apm-data/src/main/resources/component-templates/apm@mappings.yaml [hunk 3]
```diff
@@ -17,3 +17,7 @@ template:
         type: constant_keyword
       data_stream.namespace:
         type: constant_keyword
+      event.ingested:
+        type: date
+        format: "strict_date_time_no_millis||strict_date_optional_time||epoch_millis"
+        ignore_malformed: false
```
Comment (inge4pres) [inline x-pack/plugin/apm-data/src/main/resources/component-templates/apm@mappings.yaml:23]: do we need to redefine this, when it's already part of ECS, so imported through `ecs@mappings`?

File: x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml (hunk 2 above)
Comment (endorama) [reply x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:28]: What would be the reason not to?

File: x-pack/plugin/apm-data/src/main/resources/component-templates/apm@mappings.yaml (hunk 3 above)
Comment (axw) [reply x-pack/plugin/apm-data/src/main/resources/component-templates/apm@mappings.yaml:23]: I don't think we have to, unless you want to use an indexed runtime field like we have in the trace.sampled data stream: https://github.com/elastic/elasticsearch/blob/4e8e9731567d66ed4cb7c3a2c7ed5af744e32ae7/x-pack/plugin/apm-data/src/main/resources/index-templates/traces-apm.sampled%40template.yaml#L33-L36

**Comment by axw** [comment]
@inge4pres could you please add some test cases to https://github.com/elastic/elasticsearch/tree/main/x-pack/plugin/apm-data/src/yamlRestTest/resources/rest-api-spec/test? See https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/apm-data/README.md#integration-testing for how to run them.

File: x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml (hunk 1 above)
Comment (inge4pres) [reply x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:36]: Yes I would like to preserve it if it's already set

File: x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml (hunk 2 above)
Comment (inge4pres) [reply x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:28]: Good point, I don't see any, but I also have a very limited understanding of what event.ingested is used for...

File: x-pack/plugin/apm-data/src/main/resources/component-templates/apm@mappings.yaml (hunk 3 above)
Comment (inge4pres) [reply x-pack/plugin/apm-data/src/main/resources/component-templates/apm@mappings.yaml:23]: ok thanks i'll remove it

File: x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml (hunk 2 above)
Comment (endorama) [reply x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:28]: I would choose based on what we were doing before removing it. As it was set by the Fleet pipeline truncating below the second I would keep this behavior. It's the least surprise for users.

File: x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml (hunk 1 above)
Comment (endorama) [reply x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:36]: I tested 2 solutions for this:
```painless
if (!ctx.event.containsKey('ingested')) {
  ctx.event.ingested = metadata().now.withNano(0).format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
}
```

and
```painless
if (ctx.event?.ingested == null) {
  ctx.event.ingested = metadata().now.withNano(0).format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
}
```

Both worked when added to the script in preventing the pipeline to override the field.

Comment (inge4pres) [reply x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:36]: Nice thanks!
I'll incorporate those, as I think it's reasonable to preserve the field when already set. 🙏🏼

**Comment by inge4pres** [comment]
@axw @endorama when looking up the ingest pipeline syntax, I found this
https://www.elastic.co/guide/en/elasticsearch/reference/current/ingest.html#access-ingest-metadata

What i don't understand is if this is able to preserve the original field while truncating to second resolution (which is what Fleet does).
Should we just reset the field with the time it is ingested? I can't seem to find a good way of retaining the field but reducing its timestamp resolution.

File: x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml (hunk 1 above)
Comment (axw) [reply x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml]: The Fleet ingest does not do that, and I don't think we should be diverging. Is there a reason we can't or shouldn't use the exact same script processor as Fleet?

https://github.com/elastic/kibana/blob/e103a253d9b756605dbeb92955ff517597055970/x-pack/plugins/fleet/server/constants/fleet_es_assets.ts#L140-L149

**Comment by axw** [comment]
> Should we just reset the field with the time it is ingested? I can't seem to find a good way of retaining the field but reducing its timestamp resolution.

I don't think we should be trying to retain it, just replace it if it's there. That's what the Fleet pipeline does. The meaning of this field is the time at which it was ingested by Elasticsearch, so it doesn't really make sense to keep any original value.

**Comment by felixbarny** [comment]
Should this target 8.x rather than main?

**Comment by inge4pres** [comment]
> Should this target 8.x rather than main?

Yes, updated 🙏🏼

**Review by lahsivjar — COMMENTED** [review]
Should we remove the explicit `event.ingested` field we are setting for traces-apm.sampled now that we are adding it to all APM signals?

https://github.com/elastic/elasticsearch/blob/9c294b29ba18588ea97e006f1ec479cc32a2885d/x-pack/plugin/apm-data/src/main/resources/index-templates/traces-apm.sampled%40template.yaml#L33-L36

**Comment by inge4pres** [comment]
> Should we remove the explicit `event.ingested` field we are setting for traces-apm.sampled now that we are adding it to all APM signals?

sgmt, removed in last commit

File: x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml [hunk 4]
```diff
@@ -24,3 +24,13 @@ processors:
     target_field: user_agent
     ignore_failure: true
     ignore_missing: true
+- script:
```
Comment (endorama) [inline x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:27]: @inge4pres This pipelines is used for all data streams, metrics included, but you added tests only for logs and traces. Should an additional test for metrics data streams be added?

Comment (inge4pres) [reply x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:27]: Ok I can do that 👍🏼

**Comment by inge4pres** [comment]
There is a CI failure unrelated with changes in this PR 

```
XPackRestIT > test {p0=migrate/10_reindex/Test Reindex With Existing Data Stream} FAILED
    java.lang.AssertionError: Failure at [migrate/10_reindex:60]: got unexpected warning header [
    	299 Elasticsearch-8.18.0-c19ca02cb1ba257721b52fcf5d9d70b0f059b1d7 "index template [my-template1] has index patterns [my-data-stream*] matching patterns from existing older templates [global] with patterns (global => [*]); this template [my-template1] will take precedence during new index creation"
    ]
        at __randomizedtesting.SeedInfo.seed([5B807D19B4B93F14:D3D442C31A4552EC]:0)
        at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:611)
        at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.test(ESClientYamlSuiteTestCase.java:554)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
        at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
        at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
        at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
        at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
        at org.elasticsearch.test.cluster.local.DefaultLocalElasticsearchCluster$1.evaluate(DefaultLocalElasticsearchCluster.java:48)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
        at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
        at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
        at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
        at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
        at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
        at java.base/java.lang.Thread.run(Thread.java:1575)
        Caused by:
        java.lang.AssertionError: got unexpected warning header [
        	299 Elasticsearch-8.18.0-c19ca02cb1ba257721b52fcf5d9d70b0f059b1d7 "index template [my-template1] has index patterns [my-data-stream*] matching patterns from existing older templates [global] with patterns (global => [*]); this template [my-template1] will take precedence during new index creation"
        ]
            at org.junit.Assert.fail(Assert.java:89)
            at org.elasticsearch.test.rest.yaml.section.DoSection.checkWarningHeaders(DoSection.java:525)
            at org.elasticsearch.test.rest.yaml.section.DoSection.execute(DoSection.java:386)
            at org.elasticsearch.test.rest.yaml.ESClientYamlSuiteTestCase.executeSection(ESClientYamlSuiteTestCase.java:591)
            ... 41 more
```

Should we just update 8.x and try again?

File: x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml (hunk 4 above)
Comment (inge4pres) [reply x-pack/plugin/apm-data/src/main/resources/ingest-pipelines/apm@pipeline.yaml:27]: Added in last commit

**Comment by inge4pres** [comment]
@reviewers please review again 🙏🏼

**Review by endorama — APPROVED** [review]

**Review by lahsivjar — APPROVED** [review]
LGTM!

**Comment by inge4pres** [comment]
Should i expect backport PRs to be automatically created?
Target branches will be 8.16 and 8.17, but i don't see any

**Comment by felixbarny** [comment]
Due to the `auto-backport` label, I would have expected back ports to happen. Maybe the reason why they didn't get created automatically was that this is targeting 8.x instead of main.

You can create them manually using https://github.com/sorenlouv/backport

**Comment by inge4pres** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.17|[](https://github.com/elastic/elasticsearch/pull/119208)|
|✅|8.16|[](https://github.com/elastic/elasticsearch/pull/119209)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**